x-pack/plugins/ml/public/url_generator.ts

```diff
@@ -64,41 +48,280 @@ export class MlUrlGenerator implements UrlGeneratorsDefinition<typeof ML_APP_URL
+      case ML_PAGES.ANOMALY_DETECTION_JOBS_MANAGE:
+        return this.createAnomalyDetectionJobManagementUrl(
+          mlUrlGeneratorState as AnomalyDetectionUrlState
+        );
+      case ML_PAGES.ANOMALY_EXPLORER:
+        return this.createExplorerUrl(mlUrlGeneratorState as ExplorerUrlState);
+      case ML_PAGES.ANOMALY_DETECTION_CREATE_JOB_SELECT_TYPE:
+        return this.createAnomalyDetectionCreatJobSelectType(
+          mlUrlGeneratorState as MlGenericUrlState
+        );
+      case ML_PAGES.SINGLE_METRIC_VIEWER:
+        return this.createSingleMetricViewerUrl(mlUrlGeneratorState as TimeSeriesExplorerUrlState);
+      case ML_PAGES.DATA_FRAME_ANALYTICS_JOBS_MANAGE:
+        return this.createDataframeAnalyticsUrl(mlUrlGeneratorState as DataFrameAnalyticsUrlState);
+      case ML_PAGES.DATA_FRAME_ANALYTICS_EXPLORATION:
+        return this.createDataframeAnalyticsExplorationUrl(
+          mlUrlGeneratorState as DataFrameAnalyticsExplorationUrlState
+        );
+      case ML_PAGES.DATA_VISUALIZER:
+        return this.createDataVisualizerUrl(mlUrlGeneratorState as DataVisualizerUrlState);
+      case ML_PAGES.DATA_VISUALIZER_FILE:
+        return this.createDataVisualizerUrl(mlUrlGeneratorState as DataVisualizerUrlState);
+      case ML_PAGES.DATA_VISUALIZER_INDEX_SELECT:
+        return this.createDataVisualizerUrl(mlUrlGeneratorState as DataVisualizerUrlState);
+      case ML_PAGES.DATA_VISUALIZER_INDEX_VIEWER:
+        return this.createIndexDataVisualizerUrl(mlUrlGeneratorState as MlGenericUrlState);
+      default:
+        throw new Error('Page type is not provided or unknown');
     }
-    throw new Error('Page type is not provided or unknown');
   };
 
+  private static extractParams<UrlState>(urlState: UrlState) {
+    // page should be guaranteed to exist here but <UrlState> is unknown
+    // @ts-ignore
+    const { page, ...params } = urlState;
+    return { page, params };
+  }
+  /**
+   * Creates URL to the Anomaly Detection Job management page
+   */
+  private createAnomalyDetectionJobManagementUrl(
+    mlUrlGeneratorState: AnomalyDetectionUrlState
+  ): string {
+    const { params } = MlUrlGenerator.extractParams<AnomalyDetectionUrlState>(mlUrlGeneratorState);
+    let url = `${this.params.appBasePath}/${ML_PAGES.ANOMALY_DETECTION_JOBS_MANAGE}`;
+    if (isEmpty(params)) {
+      return url;
+    }
+    const { jobId, groupIds } = params;
+    const queryState: AnomalyDetectionQueryState = {
+      jobId,
+      groupIds,
+    };
+
+    url = setStateToKbnUrl<AnomalyDetectionQueryState>(
+      'mlManagement',
+      queryState,
+      { useHash: false, storeInHashQuery: false },
+      url
+    );
+    return url;
+  }
+
+  private createAnomalyDetectionCreatJobSelectType(mlGenericUrlState: MlGenericUrlState): string {
+    return this.createIndexBasedMlUrl(mlGenericUrlState);
+  }
+
   /**
    * Creates URL to the Anomaly Explorer page
    */
   private createExplorerUrl({
+    refreshInterval,
```

Comment:
this file is now very large. would it be possible/convenient to break it into multiple files? one per page perhaps.
It might make it easier to manage in the future.

Reply:
Refactor to separate files here https://github.com/elastic/kibana/pull/75696/commits/4d6a5a0e30e0046121d1febced6b0399da72b25a